### PR TITLE
Improve set statements

### PR DESCRIPTION
**Pull request by yuananf**
This PR adds support for following kind of set stataments:
```
set autocommit = true
set autocommit = on
set autocommit = `on`
set autocommit = off
set @@session.autocommit = 1
set @@session.`autocommit` = 1
set @@session."autocommit" = 1
set @@session.'autocommit' = 1
```

Tests are also added.

### CODE REVIEW

**Comment by jvaidya** [comment]
Thanks for tackling this. This will help with the preamble queries from
Python/Connector.

Could you please add test for:

set session @@global.autocommit = 'OFF';

Also, this may be a little out of the scope of this CR, but optinally, if
you can make this work, that would be great too:

set names utf8 collate 'foo'

File: go/vt/sqlparser/analyzer.go [hunk 1]
```diff
@@ -278,28 +284,57 @@ func ExtractSetValues(sql string) (keyValues map[string]interface{}, scope strin
 	if !ok {
 		return nil, "", fmt.Errorf("ast did not yield *sqlparser.Set: %T", stmt)
 	}
-	result := make(map[string]interface{})
+	result := make(map[SetKey]interface{})
 	for _, expr := range setStmt.Exprs {
+		scope := SessionStr
 		key := expr.Name.Lowered()
+		if strings.HasPrefix(key, "@@") {
```
[the remark this answers is not in the document]
Comment (yuananf) [reply go/vt/sqlparser/analyzer.go]: done

File: go/vt/sqlparser/analyzer.go [hunk 2]
```diff
@@ -278,28 +284,57 @@ func ExtractSetValues(sql string) (keyValues map[string]interface{}, scope strin
...
+			if strings.HasPrefix(key, "@@global.") {
```
[the remark this answers is not in the document]
Comment (yuananf) [reply go/vt/sqlparser/analyzer.go]: done

File: go/vt/sqlparser/analyzer_test.go [hunk 3]
```diff
@@ -375,38 +375,59 @@ func TestExtractSetValues(t *testing.T) {
...
+		out: map[SetKey]interface{}{{Key: "autocommit", Scope: "session"}: int64(1)},
+	}, {
+		sql: "set @@session.\"autocommit\"=1",
+		out: map[SetKey]interface{}{{Key: "autocommit", Scope: "session"}: int64(1)},
+	}, {
+		sql: "set @@session.'\"autocommit'=1",
+		out: map[SetKey]interface{}{{Key: "\"autocommit", Scope: "session"}: int64(1)},
+	}, {
+		sql: "set @@session.`autocommit'`=1",
+		out: map[SetKey]interface{}{{Key: "autocommit'", Scope: "session"}: int64(1)},
```
[the remark this answers is not in the document]
Comment (yuananf) [reply go/vt/sqlparser/analyzer_test.go]: tests added

File: go/vt/vtgate/executor.go [hunk 4]
```diff
@@ -526,17 +526,32 @@ func (e *Executor) handleSet(ctx context.Context, safeSession *SafeSession, sql
...
+		switch k.Key {
 		case "autocommit":
-			val, ok := v.(int64)
-			if !ok {
+			var val int64
+			switch v := v.(type) {
+			case int64:
+				val = v
+			case string:
+				if v == "on" {
```
[the remark this answers is not in the document]
Comment (yuananf) [reply go/vt/vtgate/executor.go]: actually on is a keyword and I set this lower case value in parser

File: go/vt/sqlparser/token.go [hunk 5]
```diff
@@ -475,7 +475,11 @@ func (tkn *Tokenizer) Scan() (int, []byte) {
 				return tkn.scanBitLiteral()
 			}
 		}
-		return tkn.scanIdentifier(byte(ch))
+		isDbSystemVariable := false
+		if ch == '@' && tkn.lastChar == '@' {
+			isDbSystemVariable = true
+		}
+		return tkn.scanIdentifier(byte(ch), isDbSystemVariable)
```
[the remark this answers is not in the document]
Comment (yuananf) [reply go/vt/sqlparser/token.go]: done

File: go/vt/sqlparser/ast.go [hunk 6]
```diff
@@ -3337,10 +3337,18 @@ func (node *TableIdent) UnmarshalJSON(b []byte) error {
...
+	isDbSystemVariable := false
+	if len(original) > 1 && original[:2] == "@@" {
+		isDbSystemVariable = true
+	}
+
 	for i, c := range original {
 		if !isLetter(uint16(c)) {
-			if i == 0 || !isDigit(uint16(c)) {
-				goto mustEscape
+			if !isDbSystemVariable ||
```
[the remark this answers is not in the document]
Comment (yuananf) [reply go/vt/sqlparser/ast.go]: done

**Comment by yuananf** [comment]
@jvaidya this sql is illegal, since it mixed using variable scope
```set session @@global.autocommit = 'OFF';```
and will throw an error.
see the test for ```set global @@session.client_found_rows = 1```

As for ```set names utf8 collate 'foo'```, I may make another PR about it, this PR already contains too many changes.

**Pull request merged**